src/test/unit_tests/test_s3select.js

```diff
@@ -0,0 +1,205 @@
+        let select = await this.s3select.write(chunk);
+        if (select) {
+            this.push(select.select);
+        }
+        return cb();
+    }
+
+    async _flush(cb) {
+        //console.log("in flush");
+        let select = await this.s3select.flush();
+        if (select) {
+            this.push(select.select);
+        }
+        return cb();
+    }
+}
+
+class CollectChunks extends stream.Writable {
+
+    chunks = [];
+
+    _write(chunk, enc, next) {
+        //console.log("got chunk");
+        this.chunks.push(chunk);
+        next();
+    }
+
+}
+
+/* DummyTransform is used a middle stream between
+input and output streams, like S3SelectStream.
+It is used to check that output stream is properly
+closed when input stream has errors*/
+class DummyTransform extends Transform {
+
+    constructor(writable) {
+        super({});
+        this.writable = writable;
+    }
+
+    _transform(chunk, enconding, cb) {
+        this.writable.write(chunk);
+        return cb();
+    }
+
+    _flush(cb) {
+        if (this.writable.flush) {
+            this.writable.flush();
+        }
+        return cb();
+    }
+
+}
+
+async function run_sql(args, input_str) {
+    const s3select = new S3SelectStream(args);
+    const input_stream = stream.Readable.from([input_str]);
+    const collect = new CollectChunks();
+    await stream.promises.pipeline(input_stream, s3select, collect);
+    return Buffer.concat(collect.chunks).toString();
+}
+
+async function simulate_async_loop(writable) {
+    writable.write("this is an ");
+    await new Promise(resolve => {
+        writable.write("input string.", null, () => {
+            resolve();
+        });
+    });
+    assert(writable.emit('error', new Error("I am an error event in an async loop.")));
+}
+
+
+mocha.describe('s3select', function() {
+
+    mocha.it('should create ok', function() {
+        const select_args = {
+            //query casts second column into integers, and sums them.
+            query: "select sum(int(_2)) from stdin;",
+            input_format: "CSV",
+            input_serialization_format: {"FieldDelimiter": ",", "RecordDelimiter": "\n"},
+            records_header_buf: s3select_utils.S3SelectStream.records_message_headers
+        };
+        const s3select = new s3select_utils.S3SelectStream(select_args);
+        s3select.destroy();
+    });
+
+    mocha.it('select star', async function() {
+        const select_args = {
+            query: "select * from stdin;",
+            input_format: "CSV",
+            input_serialization_format: {"FieldDelimiter": ",", "RecordDelimiter": "\n"},
+            records_header_buf: s3select_utils.S3SelectStream.records_message_headers
+        };
+        const output = await run_sql(select_args, csv_small_str);
+        assert.strictEqual(output, csv_small_str);
+    });
+
+    mocha.it('select sum with where', async function() {
+        const select_args = {
+            //sum every third number in column 2
```

Comment:
The comment is good, but not accurate, please add brackets.
something like:
`(takes the numbers from rows number 1, 4 and 7 from column 2: 858 + 636 + 60 = 1,554).`